src/sampling.rs

```diff
@@ -49,22 +90,68 @@ impl<F> Sampler<F> {
         let normal = Normal::new(0.0, 1.0).unwrap();
 
         // Random steps in the distribution N(0, I)
-        let z = (0..self.population_size)
-            .map(|_| {
-                DVector::from_iterator(
-                    self.dim,
-                    (0..self.dim).map(|_| normal.sample(&mut self.rng)),
-                )
-            })
-            .collect::<Vec<_>>();
-        let transform = |zk| state.cov_transform() * zk;
-        let y = if parallel_update {
-            z.into_par_iter().map(transform).collect()
-        } else {
-            z.into_iter().map(transform).collect()
+        let mut sample = |n: usize, constraints: Option<&dyn Constraints>| {
+            let z = (0..n)
+                .map(|_| {
+                    DVector::from_iterator(
+                        self.dim,
+                        (0..self.dim).map(|_| normal.sample(&mut self.rng)),
+                    )
+                })
+                .collect::<Vec<_>>();
+            let transform = |zk| state.cov_transform() * zk;
+
+            match constraints {
+                Some(constraints) => {
+                    let ok_constraints = |yk: &DVector<f64>| {
+                        let point = to_point(&yk, state.mean(), state.sigma());
+                        constraints.meets_constraints(&point)
+                    };
+
+                    if parallel_update {
+                        z.into_par_iter()
+                            .map(transform)
+                            .filter(ok_constraints)
+                            .collect()
+                    } else {
+                        z.into_iter()
+                            .map(transform)
+                            .filter(ok_constraints)
+                            .collect()
+                    }
+                }
+                None => {
+                    if parallel_update {
+                        z.into_par_iter().map(transform).collect()
+                    } else {
+                        z.into_iter().map(transform).collect()
+                    }
+                }
+            }
```

Comment:
The `Some` check can be moved into `ok_constraints` so `filter` can be applied unconditionally. Something like this:
```rust
let ok_constraints = |yk: &DVector<f64>| {
    match constraints {
        Some(c) => c.meets_constraints(&to_point(&yk, state.mean(), state.sigma())),
        None => true,
    }
};
```

Reply:
I wanted to keep this out of the filter for performance reasons; the check does not need to run on every iteration. But maybe that's premature optimization - I haven't benchmarked this. I'll check with my current use case and move it into the lambda if the performance is not changed.